revocation/internal/crl/crl.go

```diff
@@ -73,37 +71,67 @@ func CertCheckStatus(ctx context.Context, cert, issuer *x509.Certificate, opts C
 			Result: result.ResultNonRevokable,
 			ServerResults: []*result.ServerResult{{
 				RevocationMethod: result.RevocationMethodCRL,
+				Error:            errors.New("CRL is not supported"),
 				Result:           result.ResultNonRevokable,
 			}},
 			RevocationMethod: result.RevocationMethodCRL,
 		}
 	}
 
+	var (
+		serverResults = make([]*result.ServerResult, 0, len(cert.CRLDistributionPoints))
+		lastErr       error
+		crlURL        string
+	)
+
+	cachedFetcher, err := fetcher.NewCachedFetcher(opts.HTTPClient, opts.CacheClient)
+	if err != nil {
+		return &result.CertRevocationResult{
+			Result: result.ResultUnknown,
+			ServerResults: []*result.ServerResult{{
+				Result:           result.ResultUnknown,
+				Error:            err,
+				RevocationMethod: result.RevocationMethodCRL,
+			}},
+			RevocationMethod: result.RevocationMethodCRL,
+		}
+	}
+
 	// The CRLDistributionPoints contains the URIs of all the CRL distribution
 	// points. Since it does not distinguish the reason field, it needs to check
 	// all the URIs to avoid missing any partial CRLs.
 	//
 	// For the majority of the certificates, there is only one CRL distribution
 	// point with one CRL URI, which will be cached, so checking all the URIs is
 	// not a performance issue.
-	var (
-		serverResults = make([]*result.ServerResult, 0, len(cert.CRLDistributionPoints))
-		lastErr       error
-		crlURL        string
-	)
 	for _, crlURL = range cert.CRLDistributionPoints {
-		baseCRL, err := download(ctx, crlURL, opts.HTTPClient)
+		bundle, fromCache, err := cachedFetcher.Fetch(ctx, crlURL)
 		if err != nil {
 			lastErr = fmt.Errorf("failed to download CRL from %s: %w", crlURL, err)
 			break
 		}
 
-		if err = validate(baseCRL, issuer); err != nil {
-			lastErr = fmt.Errorf("failed to validate CRL from %s: %w", crlURL, err)
-			break
+		if err = validate(bundle.BaseCRL, issuer); err != nil {
+			if fromCache {
+				// the CRL may be stale, try to download again
+				bundle, err = cachedFetcher.Download(ctx, crlURL)
+				if err != nil {
+					lastErr = fmt.Errorf("failed to download CRL from %s: %w", crlURL, err)
+					break
+				}
+
+				if err = validate(bundle.BaseCRL, issuer); err != nil {
+					lastErr = fmt.Errorf("failed to validate CRL from %s: %w", crlURL, err)
+					break
+				}
+			} else {
+				// the CRL is fresh, but it is invalid
+				lastErr = fmt.Errorf("failed to validate CRL from %s: %w", crlURL, err)
+				break
+			}
```

Comment:
Since the cache is transparent to the caller, it is unnecessary to consider things like `fromeCache`.

Reply:
Updated.